source/validate_type.cpp

```diff
@@ -0,0 +1,202 @@
+    case SPV_TYPE_CATEGORY_TYPE_VECTOR:         ret = "TypeVector"; break;
+    case SPV_TYPE_CATEGORY_TYPE_MATRIX:         ret = "TypeMatrix"; break;
+    case SPV_TYPE_CATEGORY_TYPE_ARRAY:          ret = "TypeArray"; break;
+    case SPV_TYPE_CATEGORY_TYPE_RUNTIMEARRAY:   ret = "TypeRuntimeArray"; break;
+    case SPV_TYPE_CATEGORY_TYPE_STRUCT:         ret = "TypeStructure"; break;
+    case SPV_TYPE_CATEGORY_TYPE_SAMPLEDIMAGE:   ret = "TypeSampledImage"; break;
+    case SPV_TYPE_CATEGORY_TYPE_FUNCTION:       ret = "TypeFunction"; break;
+    case SPV_TYPE_CATEGORY_TYPE_OPAQUE:         ret = "TypeOpaque"; break;
+    case SPV_TYPE_CATEGORY_TYPE_QUEUE:          ret = "TypeQueue"; break;
+    case SPV_TYPE_CATEGORY_TYPE_POINTER:        ret = "TypePointer"; break;
+    case SPV_TYPE_CATEGORY_TYPE_EVENT:          ret = "TypeEvent"; break;
+    case SPV_TYPE_CATEGORY_TYPE_DEVICEEVENT:    ret = "TypeDeviceEvent"; break;
+    case SPV_TYPE_CATEGORY_TYPE_RESERVEID:      ret = "TypeReserveId"; break;
+    case SPV_TYPE_CATEGORY_TYPE_PIPE:           ret = "TypePipe"; break;
+    case SPV_TYPE_CATEGORY_TYPE_FORWARDPOINTER: ret = "TypeForwardPointer"; break;
+    case SPV_TYPE_CATEGORY_NUMERICAL:           ret = "TypeNumerical"; break;
+    case SPV_TYPE_CATEGORY_SCALAR:              ret = "TypeScalar"; break;
+    case SPV_TYPE_CATEGORY_AGGREGATE:           ret = "TypeAggregate"; break;
+    case SPV_TYPE_CATEGORY_COMPOSITE:           ret = "TypeComposite"; break;
+    case SPV_TYPE_CATEGORY_CONCRETE:            ret = "TypeConcrete"; break;
+    case SPV_TYPE_CATEGORY_ABSTRACT:            ret = "TypeAbstract"; break;
+    case SPV_TYPE_CATEGORY_OPAQUE:              ret = "TypeOpaque"; break;
+    case SPV_TYPE_CATEGORY_ANY:                 ret = "TypeCategoryAny"; break;
+    case SPV_TYPE_CATEGORY_NONE:                ret = "NONE"; break;
+  }
+  return ret;
+}
+
+// clang-format off
+// Returns a string representation of a vector of type category flags
+string TypeFlagsToString(const vector<spv_type_category_t>& flags) {
+  string out = "{";
+  for (auto flag : flags) {
+    out += " " + TypeFlagToString(flag);
+  }
+  return out + " }";
+}
+// clang-format on
+
+struct SpvOpHash {
+  size_t operator()(SpvOp op) const {
+    return std::hash<uint32_t>()(static_cast<uint32_t>(op));
+  }
+};
+
+// TODO(umar): This representation for the type descriptors is not complete
+// and I don't think it is robust enough to handle everything in the standard.
+// TODO(umar): Support capability dependant checks
+// TODO(umar): Support decoration dependant checks
+// TODO(umar): Support type width and sign dependant checks
+struct spv_type_desc_t {
+  const SpvOp opcode;
+  const vector<spv_type_category_t> result_type;
+};
+
+// NOTE: This will be generated at build time for a specific version
+// of spirv. Ideally this should be generated by the spec and added
+// to the build script as an external dependency.
```

Comment:
👍  Overall I think this is a good approach. But we should definitely make it data-driven.